src/platforms/stlinkv3/usb_dwc_common.c

```diff
@@ -0,0 +1,480 @@
+	 * for the blackmagic use case, because the gdb protocol is
+	 * packetized, and the single zero byte will be silently discarded
+	 * by gdb, but this approach is incorrect in general.
+	 *
+	 * For the time being, return zero here in case of error. */
+	/* Return if endpoint is already enabled, which means a packet
+	 * transfer is still in progress. */
+	if (REBASE(OTG_DIEPCTL(addr)) & OTG_DIEPCTL0_EPENA)
+		return 0;
+	if (REBASE(OTG_DTXFSTS(addr) & 0xffff) < (unsigned)((len + 3) >> 2))
+		return 0;
+
+	/* Enable endpoint for transmission. */
+	REBASE(OTG_DIEPTSIZ(addr)) = OTG_DIEPSIZ0_PKTCNT | len;
+
+	/* WARNING: it is not explicitly stated in the ST documentation,
+	 * but the usb core fifo memory read/write accesses should not
+	 * be interrupted by usb core fifo memory write/read accesses.
+	 *
+	 * For example, this function can be run from within the usb interrupt
+	 * context, and also from outside of the usb interrupt context.
+	 * When this function executes outside of the usb interrupt context,
+	 * if it gets interrupted by the usb interrupt while it writes to
+	 * the usb core fifo memory, and from within the usb
+	 * interrupt the usb core fifo memory is read, then when control returns
+	 * to this function, the usb core fifo memory write accesses can not
+	 * simply continue, this will result in a transaction error on the
+	 * usb line.
+	 *
+	 * In order to avoid such situation, the usb interrupt is masked here
+	 * prior to writing the data to the usb core fifo memory.
+	 */
+	uint32_t saved_interrupt_mask = REBASE(OTG_GINTMSK);
+	REBASE(OTG_GINTMSK) = 0;
+	REBASE(OTG_DIEPCTL(addr)) |= OTG_DIEPCTL0_EPENA | OTG_DIEPCTL0_CNAK;
+
+	/* Copy buffer to endpoint FIFO, note - memcpy does not work.
+	 * ARMv7M supports non-word-aligned accesses, ARMv6M does not. */
+#if defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7EM__)
+	for (i = len; i > 0; i -= 4) {
+		REBASE(OTG_FIFO(addr)) = *buf32++;
+	}
+#endif /* defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7EM__) */
+
+#if defined(__ARM_ARCH_6M__)
+	/* Take care of word-aligned and non-word-aligned buffers */
+	if (((uint32_t)buf8 & 0x3) == 0) {
+		for (i = len; i > 0; i -= 4) {
+			REBASE(OTG_FIFO(addr)) = *buf32++;
+		}
+	} else {
+		for (i = len; i > 0; i -= 4) {
+			memcpy(&word32, buf8, 4);
+			REBASE(OTG_FIFO(addr)) = word32;
+			buf8 += 4;
+		}
+	}
+#endif /* defined(__ARM_ARCH_6M__) */
+
+	REBASE(OTG_GINTMSK) = saved_interrupt_mask;
+	return len;
+}
+
+uint16_t dwc_ep_read_packet(usbd_device *usbd_dev, uint8_t addr, void *buf, uint16_t len)
+{
+	(void)usbd_dev;
+	struct incoming_packet *packet = stashed_packets + addr;
+	if (!packet->is_packet_present)
+		return 0;
+	len = MIN(len, packet->packet_length);
+	packet->is_packet_present = false;
+	memcpy(buf, packet->packet_data, len);
+	return len;
+}
+
+/* TODO: this does not currently handle the case for __ARM_ARCH_6M__ */
+void dwc_ep_read_packet_internal(usbd_device *usbd_dev, int ep)
+{
+	int i;
```

Comment:
C89 decls

`i` should be declared just before the loop it's for given the loop condition followed by checking it after the loop